### PR TITLE
feat(web): GitLab MR Review Agent

### PR DESCRIPTION
**Pull request by fatmcgav**
Adds support for the AI Review Agent to review GitLab Merge Requests, mirroring the existing GitHub PR review functionality. Also fixes several bugs discovered during implementation and improves the shared review pipeline.

---

 ## New files

 ### `packages/web/src/features/agents/review-agent/nodes/gitlabMrParser.ts`
Parses a GitLab MR webhook payload into the shared `sourcebot_pr_payload` format. Calls `MergeRequests.show()` and `MergeRequests.allDiffs()` in parallel — the API response is used for `title`, `description`, `sha`, and `diff_refs` (which can be absent in webhook payloads for `update` action events), while per-file diffs are parsed using the existing `parse-diff` library.

 ### `packages/web/src/features/agents/review-agent/nodes/gitlabPushMrReviews.ts`
Posts review comments back to GitLab using `MergeRequestDiscussions.create()` with a position object carrying `base_sha`, `head_sha`, and `start_sha`. Falls back to `MergeRequestNotes.create()` (a general MR note) if the inline comment is rejected by the API (e.g. the line is not within the diff), ensuring reviews are always surfaced even when precise positioning fails.

 ### Test files (4 new files, 34 tests total)
- `githubPrParser.test.ts` — diff parsing and metadata mapping for the GitHub parser
- `githubPushPrReviews.test.ts` — single-line vs multi-line comment parameters, error resilience
- `gitlabMrParser.test.ts` — API call arguments, metadata mapping, diff parsing, edge cases (empty diffs, nested groups, null description, API failures)
- `gitlabPushMrReviews.test.ts` — inline comment posting, fallback behaviour, missing `diff_refs` guard, multi-file iteration

---

 ## Modified files

 ### `packages/web/src/features/agents/review-agent/types.ts`
- Added `sourcebot_diff_refs` schema/type (`base_sha`, `head_sha`, `start_sha`) and an optional `diff_refs` field on `sourcebot_pr_payload`
- Added `GitLabMergeRequestPayload` and `GitLabNotePayload` interfaces for webhook event typing

 ### `packages/web/src/features/agents/review-agent/app.ts`
- Added `processGitLabMergeRequest()` function mirroring `processGitHubPullRequest()`: sets up logging, runs the GitLab parser, generates reviews via the shared LLM pipeline, and pushes results
- Removed stale `OPENAI_API_KEY` guards (model availability is now enforced inside `invokeDiffReviewLlm`)

 ### `packages/web/src/features/agents/review-agent/nodes/invokeDiffReviewLlm.ts`
**Replaces the hardcoded OpenAI client** with the Vercel AI SDK's `generateText` and the shared `getAISDKLanguageModelAndOptions` / `getConfiguredLanguageModels` utilities from `chat/utils.server.ts`. The review agent now uses whichever language model is configured in `config.json`, supporting all providers (Anthropic, Bedrock, Azure, etc.).

- `REVIEW_AGENT_MODEL` env var (matched against `displayName`) selects a specific model when multiple are configured; falls back to `models[0]` with a warning if the name is not found
- Prompt is passed via the `system` parameter with a `"Review the code changes."` user turn, satisfying providers (e.g. Bedrock/Anthropic) that require conversations to begin with a user message

 ### `packages/web/src/features/agents/review-agent/nodes/fetchFileContent.ts`
**Fixes "Not authenticated" error** when the review agent calls `getFileSource`. The original implementation used `withOptionalAuth`, which reads a session cookie — absent in webhook handlers. Now calls `getFileSourceForRepo` directly with `__unsafePrisma` and the single-tenant org, bypassing the session-based auth layer. The webhook handler has already authenticated the request via its own mechanism (GitHub App signature / GitLab token).

 ### `packages/web/src/features/git/getFileSourceApi.ts`
- Extracted the core repo-lookup + git + language-detection logic into a new exported `getFileSourceForRepo({ path, repo, ref }, { org, prisma })` function
- `getFileSource` now handles auth and audit logging then delegates to `getFileSourceForRepo` — all existing callers are unchanged

 ### `packages/web/src/app/api/(server)/webhook/route.ts`
- Added GitLab webhook handling alongside the existing GitHub branch
- Verifies `x-gitlab-token` against `GITLAB_REVIEW_AGENT_WEBHOOK_SECRET`
- Handles `Merge Request Hook` events (auto-review on `open`, `update`, `reopen`) and `Note Hook` events (manual `/review` command on MR comments)
- Initialises a `Gitlab` client at module load if `GITLAB_REVIEW_AGENT_TOKEN` is set

 ### `packages/web/src/app/(app)/agents/page.tsx`
- Split the single "Review Agent" card into two separate cards: **GitHub Review Agent** and **GitLab Review Agent**, each showing its own configuration status
- Removed `OPENAI_API_KEY` from the GitHub card's required env vars (no longer applicable)

 ### `packages/web/src/app/(app)/components/navigationMenu/navigationItems.tsx` & `index.tsx`
- Added an **Agents** nav item (with `BotIcon`) between Repositories and Settings
- Visible when the user is authenticated **and** at least one agent is configured (GitHub App triple or GitLab token pair), computed in the server component and passed down as `isAgentsVisible`

 ### `packages/shared/src/env.server.ts`
Added four new environment variables:
| Variable | Purpose |
|---|---|
| `GITLAB_REVIEW_AGENT_WEBHOOK_SECRET` | Verifies the `x-gitlab-token` header on incoming webhooks |
| `GITLAB_REVIEW_AGENT_TOKEN` | Personal or project access token used for GitLab API calls |
| `GITLAB_REVIEW_AGENT_HOST` | GitLab hostname (defaults to `gitlab.com`; set for self-hosted instances) |
| `REVIEW_AGENT_MODEL` | `displayName` of the configured language model to use for reviews; falls back to the first model if unset or not matched |

 ### `packages/web/package.json`
Added `@gitbeaker/rest` dependency (already used in `packages/backend`).

---

 ## Bug fixes

| Bug | Fix |
|---|---|
| `"Not authenticated"` when fetching file content from the review agent | `fetchFileContent` now calls `getFileSourceForRepo` directly instead of `getFileSource` (which gates on session auth) |
| `"diff_refs is missing"` when posting GitLab MR reviews | `gitlabMrParser` now fetches the full MR via `MergeRequests.show()` instead of relying on the webhook payload, which omits `diff_refs` on `update` events |
| Bedrock/Anthropic rejection: `"A conversation must start with a user message"` | `invokeDiffReviewLlm` now passes the prompt via `system` + a `prompt` user turn instead of a `system`-role entry inside `messages` |
| Review agent silently used `models[0]` with no way to specify a different model | New `REVIEW_AGENT_MODEL` env var selects by `displayName` |

## Summary by CodeRabbit

* **New Features**
  * GitLab merge request support for the automated code review agent alongside GitHub
  * Separate GitHub and GitLab agents with per-agent “Configure” actions and an Agents page/link shown when configured
  * Runtime language-model selection and optional model override via env

* **Bug Fixes**
  * Removed hard OpenAI-key guard so configured models drive behavior
  * .gitignore updated

* **Tests**
  * Added comprehensive parser and posting tests for GitHub and GitLab review flows

* **Documentation**
  * Updated review-agent docs with GitLab setup, model selection, and env var guidance

* **Chores**
  * Added GitLab client dependency and updated changelog

### CODE REVIEW

**Comment by fatmcgav** [comment]
I've tested this locally against a Gitlab repo using AWS Bedrock, and got a successful review comment:
image

File: packages/web/src/features/agents/review-agent/nodes/gitlabPushMrReviews.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                    {
+                        position: {
+                            positionType: "text",
+                            baseSha: base_sha,
+                            headSha: head_sha,
+                            startSha: start_sha,
+                            newPath: fileDiffReview.filename,
+                            newLine: String(review.line_end),
+                        },
+                    },
```
[the remark this answers is not in the document]
Comment (fatmcgav) [reply packages/web/src/features/agents/review-agent/nodes/gitlabPushMrReviews.ts:38]: Clade thinks this one is incorrect: 

> The Gitbeaker API expects camelCase — positionType, baseSha, headSha, startSha, newPath, newLine — which is exactly what the current code uses. The finding is incorrect.
>
>  However, there are two genuine issues the type definition reveals:
>  1. newLine should be string (it already is — String(review.line_end) is correct per the type)
>  2. oldPath is missing — the type includes it as optional, and adding it is good practice for renamed files
>
>  The type confirms the current code is correct. No changes are needed.

**Comment by fatmcgav** [comment]
@msukkari @brendan-kellam A review on this one would be greatly appreciated, as we're keen to give it a try on a few repos internally :)

**Comment by fatmcgav** [comment]
@coderabbitai review

**Comment by aiSynergy37** [comment]
Prepared a fix for the two open CodeRabbit findings:

1) fallback `head_sha` to webhook `last_commit.id` in `gitlabMrParser.ts`
2) tighten GitLab webhook discriminators in `types.ts` (`object_kind` literals and `noteable_type` literal)

Patch is available on my fork branch: `aiSynergy37:fix/pr-1104-gitlab-webhook-schemas`
Commit: `63220b0c83a19f53f47049cd95a3db5c3f8de8c4`

If helpful, maintainers can cherry-pick directly:
`git cherry-pick 63220b0c83a19f53f47049cd95a3db5c3f8de8c4`

**Comment by fatmcgav** [comment]
@coderabbitai review

**Comment by fatmcgav** [comment]
@msukkari / @brendan-kellam I've rebased this branch based on the changes in !1134. 

Let me know if needs anything additional, or if we could get this merged and released as an experimental feature? 👍

**Review by brendan-kellam — APPROVED** [review]
LGTM thanks!

**Comment by brendan-kellam** [comment]
release build is running atm, will merge afterwards

**Comment by gaeljw** [comment]
Wow, thanks folks, I'll definitely give this a try in our projects as well! 👏👏

**Comment by fatmcgav** [comment]
> Wow, thanks folks, I'll definitely give this a try in our projects as well! 👏👏

@gaeljw I've just opened #1143 which takes things one step further to support "dynamic" agents :) 

I think I've covered most of the bases, but feel free to take a look and suggest any additions...
